dd-sdk-android-ndk/src/main/cpp/utils/signal-monitor.c

```diff
@@ -236,54 +268,22 @@ bool try_to_install_handlers() {
     if (handlers_installed) {
         return true;
     }
-
     return (configure_signal_stack() &&
-            configure_global_sigaction() &&
+            init_datadog_sigaction() &&
             start_watchdog_thread());
 }
 
-bool install_signal_handlers() {
+bool start_monitoring() {
     pthread_mutex_lock(&mutex);
     bool installed = try_to_install_handlers();
     pthread_mutex_unlock(&mutex);
     return installed;
 }
 
-void free_up_memory() {
-    if (global_sigaction != NULL) {
-        free(global_sigaction);
-    }
-    if (global_previous_sigaction != NULL) {
-        free(global_previous_sigaction);
-    }
-    if (signal_stack.ss_sp != NULL) {
-        free(signal_stack.ss_sp);
-    }
-    global_sigaction = NULL;
-    global_previous_sigaction = NULL;
-    signal_stack.ss_sp = NULL;
-}
-
-void uninstall_signal_handlers() {
+void stop_monitoring() {
     pthread_mutex_lock(&mutex);
-
-    if (!handlers_installed) {
-        pthread_mutex_unlock(&mutex);
-        return;
-    }
-
-    const size_t signals_array_size = handled_signals_size();
-    for (int i = 0; i < signals_array_size; i++) {
-        struct sigaction *prev_action = &global_previous_sigaction[i];
-        if (prev_action) {
-            const int signal = handled_signals[i].signal_value;
-            sigaction(signal, prev_action, 0);
-        }
-    }
-
+    uninstall_handlers();
     free_up_memory();
-    recordUninstallOp();
-    handlers_installed = false;
     pthread_mutex_unlock(&mutex);
```

Comment:
can it be the issue here if we try to unlock mutex which is already unlocked? because unlock can also happen in the `uninstall_handlers` which is called above.

Reply:
that's a good point. I will test this, normally it should do nothing if the lock is already unlocked but it worths the try.

Reply:
Actually with the new implementation the `unlock` is not needed anymore. Also FYI the process will crash with `Unefined behaviour` if you try to unlock an already unlocked lock 🤦 . That is good to know 👍